R/aaa-.r

```diff
@@ -12,3 +12,44 @@ NULL
 #' @keywords internal
 #' @name ggplot2-ggproto
 NULL
+
+# Fast data.frame constructor and indexing
+# No checking, recycling etc. unless asked for
+new_data_frame <- function(x = list(), n = NULL) {
+  if (length(x) != 0 && is.null(names(x))) stop("Elements must be named", call. = FALSE)
+  lengths <- vapply(x, length, integer(1))
+  if (is.null(n)) {
+    n <- if (length(x) == 0) 0 else max(lengths)
+  }
+  for (i in seq_along(x)) {
+    if (lengths[i] == n) next
+    if (lengths[i] != 1) stop("Elements must equal the number of rows or 1", call. = FALSE)
+    x[[i]] <- rep(x[[i]], n)
+  }
+
+  class(x) <- "data.frame"
```

Comment:
Might be a little cleaner to set with `structure` ?

Reply:
`structure()` is measurably slower... If the purpose is performance I think we can sacrifice a bit of cleanliness

Reply:
Oh wow, so it is. Maybe this means vctrs will need it's own `structure()` alternative.

Reply:
I was a bit surprised, honestly